core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -836,64 +837,89 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+             (includeStores || storeStatus.hasCharacteristic(Characteristic.READ_ONLY) ||
+              storeStatus.config.ignoreModifications());
    }
 
-   private boolean isLocalOnlyLoader(NonBlockingStore store) {
+   private boolean isLocalOnlyLoader(NonBlockingStore<?, ?> store) {
       if (store instanceof LocalOnlyCacheLoader) return true;
-      NonBlockingStore unwrappedStore;
+      NonBlockingStore<?, ?> unwrappedStore;
       if (store instanceof DelegatingNonBlockingStore) {
-         unwrappedStore = ((DelegatingNonBlockingStore) store).delegate();
+         unwrappedStore = ((DelegatingNonBlockingStore<?, ?>) store).delegate();
       } else {
          unwrappedStore = store;
       }
       if (unwrappedStore instanceof LocalOnlyCacheLoader) {
          return true;
       }
       if (unwrappedStore instanceof NonBlockingStoreAdapter) {
-         return ((NonBlockingStoreAdapter) unwrappedStore).getActualStore() instanceof LocalOnlyCacheLoader;
+         return ((NonBlockingStoreAdapter<?, ?>) unwrappedStore).getActualStore() instanceof LocalOnlyCacheLoader;
       }
       return false;
    }
 
    @Override
-   public CompletionStage<Long> size(Predicate<? super StoreConfiguration> predicate) {
+   public CompletionStage<Long> approximateSize(Predicate<? super StoreConfiguration> predicate, IntSet segments) {
+      if (!isEnabled()) {
+         return NonBlockingStore.SIZE_UNAVAILABLE_FUTURE;
+      }
       long stamp = acquireReadLock();
       try {
-         checkStoreAvailability();
-         if (log.isTraceEnabled()) {
-            log.tracef("Obtaining size from stores");
+         if (!isAvailable()) {
+            releaseReadLock(stamp);
+            return NonBlockingStore.SIZE_UNAVAILABLE_FUTURE;
          }
-         NonBlockingStore<?, ?> nonBlockingStore = getStoreLocked(storeStatus -> storeStatus.characteristics.contains(
-               Characteristic.BULK_READ) && predicate.test(storeStatus.config));
-         if (nonBlockingStore == null) {
+
+         // Ignore stores without BULK_READ, they don't implement approximateSize()
+         StoreStatus firstStoreStatus = getStoreStatusLocked(storeStatus -> {
+            return storeStatus.hasCharacteristic(Characteristic.BULK_READ) &&
+                   predicate.test(storeStatus.config);
+         });
+         if (firstStoreStatus == null) {
             releaseReadLock(stamp);
-            return CompletableFuture.completedFuture(-1L);
+            return NonBlockingStore.SIZE_UNAVAILABLE_FUTURE;
+         }
+
+         if (log.isTraceEnabled()) {
+            log.tracef("Obtaining approximate size from store %s", firstStoreStatus.store);
+         }
+         if (firstStoreStatus.hasCharacteristic(Characteristic.SEGMENTABLE)) {
+            return firstStoreStatus.store.approximateSize(segments)
+                                         .whenComplete((ignore, ignoreT) -> releaseReadLock(stamp));
```

Comment:
Actually just realized this line is not in the else, so the read lock isn't released :(

Reply:
I am fixing it in my PR https://github.com/infinispan/infinispan/pull/9845

Reply:
Oops, you're quite right. I wonder why we have no test failures, though...

Reply:
Assuming we just don't have a test that hits that line for w/e reason.